[WIP] Application kill switch

facuspagnuolo opened the pull request:
Follow up on #516 

Based on an offline discussion we had about the next-steps regarding the kill-switch topic, we decided to:
[x] Drop the kernel level kill-switch idea, at least for the MVP version
[x] Simplify the whole architecture to support only application level kill-switch using severities
[x] Support having one issue registry per application
[ ] Integrate with the rest of the aragonOS components
[ ] Keep investigating how an organization is able to recover (for now we are leaving on a side the idea of allowing DAOs to programmatically override the application decision of how it should behave when the kill-switch is turned on)

bingen commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    function shouldDenyCallingContract(address _contract) external returns (bool) {
+        // if the call should be denied, then deny given call
```
That sounds pretty obvious, haha ;-P Maybe first "call" should be "contract"?

bingen commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,122 @@
+        emit IssuesRegistrySet(_contract, address(_issuesRegistry));
+    }
+
+    function shouldDenyCallingContract(address _contract) external returns (bool) {
```
It would be nice to have an estimation of how much gas overload means adding the modifier that calls this to a function.

bingen commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,122 @@
+
+    function shouldDenyCallingContract(address _contract) external returns (bool) {
+        // if the call should be denied, then deny given call
+        if (isContractDenied(_contract)) {
```
What's exactly the point of this? Wouldn't it be the same to add an issue of maximum severity for the contract?

facuspagnuolo replied:
I'm thinking the case where a DAO don't have access to submit issues to the registry, then, once again you have the chance for a last call on the DAO's side

facuspagnuolo commented on contracts/kill_switch/IssuesRegistry.sol:
```diff
@@ -0,0 +1,28 @@
+        initialized();
+    }
+
+    function setSeverityFor(address entry, Severity severity) external authP(SET_ENTRY_SEVERITY_ROLE, arr(entry, msg.sender)) {
```
I had a brain fart :)

facuspagnuolo commented:
@bingen thanks for your review, I addressed all your comments in the last commit (https://github.com/aragon/aragonOS/pull/518/commits/ffff1702521eebb43f4fdfe5cdf4a27a4cac0723), feel free to take another look. Regarding your question about the idea of overriding the KillSwitch, yes, it's a wip, but I'm probably going to push a separate PR for that.

izqui commented on contracts/kernel/KernelConstants.sol:
```diff
@@ -10,10 +10,12 @@ contract KernelAppIds {
     bytes32 internal constant KERNEL_CORE_APP_ID = apmNamehash("kernel");
     bytes32 internal constant KERNEL_DEFAULT_ACL_APP_ID = apmNamehash("acl");
     bytes32 internal constant KERNEL_DEFAULT_VAULT_APP_ID = apmNamehash("vault");
+    bytes32 internal constant KERNEL_DEFAULT_KILL_SWITCH_APP_ID = apmNamehash("killSwitch");
```
Nit: App names use snake case 

```suggestion
    bytes32 internal constant KERNEL_DEFAULT_KILL_SWITCH_APP_ID = apmNamehash("kill-switch");
```

(will need to update the actual hash as well)

izqui commented on contracts/kill_switch/IssuesRegistry.sol:
```diff
@@ -0,0 +1,28 @@
+        initialized();
+    }
+
+    function setSeverityFor(address entry, Severity severity) external authP(SET_ENTRY_SEVERITY_ROLE, arr(entry)) {
```
It would be clearer if `entry` was renamed to `baseContract` or `implementation`

izqui replied:
This will allow to parametrize permissions depending on the severity being set

```suggestion
    function setSeverityFor(address entry, Severity severity) external authP(SET_ENTRY_SEVERITY_ROLE, arr(entry, uint256(severity))) {
```

izqui commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,130 @@
+
+    string constant private ERROR_ISSUES_REGISTRY_NOT_CONTRACT = "KS_ISSUES_REGISTRY_NOT_CONTRACT";
+
+    enum ContractAction { Allow, Check, Deny }
```
As we chatted offline, not all actions are useful on a per base/implementation contract basis.

With the current implementation, if checks are enabled for an app when it is upgraded the new implementation contract would need to be enabled for kill switch checks. IMO for most uses, it would be better if the `Check` action is applied per `appId`.

Similarly, the `Allow` action makes more sense to be applied for proxy addresses, which allows to bypass the kill switch check for a specific app. Marking a critical app proxy (like the root authority Voting instance in an organization) with `Allow` would be a way to prevent the organization from ever bricking (at the expense of not being protected by the kill switch)

sohkai commented on contracts/kernel/Kernel.sol:
```diff
@@ -54,6 +56,27 @@ contract Kernel is IKernel, KernelStorage, KernelAppIds, KernelNamespaceConstant
+        IKillSwitch killSwitch = IKillSwitch(newAppProxy(this, KERNEL_DEFAULT_KILL_SWITCH_APP_ID));
+        killSwitch.initialize(_issuesRegistry);
+        _setApp(KERNEL_APP_ADDR_NAMESPACE, KERNEL_DEFAULT_KILL_SWITCH_APP_ID, killSwitch);
+    }
```
To avoid coupling the Kernel too much with the KillSwitch, I think this bit of logic could be moved higher up the stack to the `DAOFactory`.

I think this would also make it more explicit that your Kernel can function without a KillSwitch, and that you can install and remove the KillSwitch as wanted.

sohkai commented on contracts/kill_switch/IKillSwitch.sol:
```diff
@@ -0,0 +1,10 @@
+
+
+contract IKillSwitch {
+    function initialize(IIssuesRegistry _issuesRegistry) external;
```
Both here and `IIssuesRegistry`, I'd prefer removing `initialize()` from the interface.

It feels like an abstraction leak to have this here; it should only be the initializer / installer of these two apps that needs to know about its initialization protocol (another good reason to move the KillSwitch's initialization into the `DAOFactory`, as that's where both the ACL and KillSwitch are "real" rather than at the Kernel, where they're abstract).

sohkai commented on contracts/kill_switch/IIssuesRegistry.sol:
```diff
@@ -0,0 +1,16 @@
+contract IIssuesRegistry {
+    enum Severity { None, Low, Mid, High, Critical }
+
+    event SeveritySet(address indexed entry, Severity severity, address sender);
```
May want to index the sender

sohkai commented on contracts/kill_switch/IssuesRegistry.sol:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    function setSeverityFor(address entry, Severity severity) external authP(SET_ENTRY_SEVERITY_ROLE, arr(entry)) {
+        issuesSeverity[entry] = severity;
```
Should we ensure that the new severity isn't lower than a previous one?

facuspagnuolo replied:
The thing is that there is not way to roll it back then, I'm thinking of a false-positive scenario

sohkai commented on contracts/kill_switch/IssuesRegistry.sol:
```diff
@@ -0,0 +1,28 @@
+        emit SeveritySet(entry, severity, msg.sender);
+    }
+
+    function isSeverityFor(address entry) public view isInitialized returns (bool) {
```
We should come up with a better name for this, perhaps `hasSeverity()` or `isAwareOfSeverity()`

sohkai commented on contracts/apps/AragonApp.sol:
```diff
@@ -31,6 +32,12 @@ contract AragonApp is AppStorage, Autopetrified, VaultRecoverable, ReentrancyGua
         _;
     }
 
+    modifier killSwitched {
```
Wondering if we should directly bundle this into `auth()` or something similar as well. It may be tedious for developers to place this on every external, state-changing call (similar to `isInitialized()`).

We may also want to find a better name, e.g. `killSwitchProtected`.

facuspagnuolo replied:
I think it may make sense but before coupling them we can keep working on other details and leave that decision for later. We will probably have a better understanding of how the whole kill-switch thing interacts with a DAO once we have finished some of the pending stuff.

OTOH, I like the new name you propose :)

bingen replied:
Yes, makes sense, although I wonder if, given the gas overhead, it could be useful to have that granularity to only apply the killSwitch to very critical functions.
Nevertheless, one could make the argue that if the function has an ACL protection is likely critical. :thinking:

facuspagnuolo commented:
@izqui @bingen @sohkai thanks for your reviews, I already addressed all your comments, and I really like how this is looking :)

OTOH, I've been exploring many different edge cases about recoverability, and unfortunately my conclusion is that we will need to keep working on further models to have a robust implementation. I filed an [issue](https://github.com/aragon/aragonOS/issues/523) sharing my thoughts on the matter, feel free to comment and leave feedback there. 
However, based on the last changes, I think we can are pretty to close to have a full MVP version of the kill-switch, we just need to decide the MVP version of the recoverability process (please follow-up in the [issue](https://github.com/aragon/aragonOS/issues/523) mentioned above).

izqui commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,142 @@
+            return true;
+        }
+
+        // if the app severity found is ignored, then allow given call
```
I would clarify that this is the part of the code where we check against the issues registry, maybe we could rename `isSeverityIgnored` to something more explicit.

facuspagnuolo replied:
I improved the comment there and rename the function to `checkSeverityIssuesFor`, WDYT?

sohkai commented on contracts/test/mocks/kill_switch/FailingKillSwitchMock.sol:
```diff
@@ -0,0 +1,12 @@
+import "../../../kill_switch/KillSwitch.sol";
+
+
+contract FailingKillSwitchMock is KillSwitch {
```
Nitpick: perhaps we can reword the name and error to be "revert" for clarity ("failing" is ambiguous as to if it doesn't work correctly by returning bad values or by reverting).

sohkai commented on test/contracts/kill_switch/issues_registry.test.js:
```diff
@@ -0,0 +1,129 @@
+      })
+
+      context('when the issues was already fixed', () => {
+        beforeEach('set medium severity', async () => {
```
```suggestion
        beforeEach('roll back severity', async () => {
```

sohkai commented on test/contracts/kill_switch/issues_registry.test.js:
```diff
@@ -0,0 +1,129 @@
+        })
+      })
+
+      context('when the issues was already fixed', () => {
```
```suggestion
      context('when the issues was a false positive', () => {
```

It's unrealistic that the issue would be fixed without updating the base implementation contract (it would have to be an dynamically linked external contract that got changed), so this might be a more realistic scenario 😄

sohkai commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,142 @@
+     */
+    function shouldDenyCallingApp(bytes32 _appId, address _base, address _instance) external returns (bool) {
+        // if the instance is allowed, then allow given call
+        if (isInstanceAllowed(_instance)) {
```
Wonder if we should a check for `_instance == address(this)` here, so that all KillSwitch instances implicitly can't turn themselves off.

facuspagnuolo replied:
Yea, I was wondering that too, but I finally decided to whitelist the kill switch instance from the factory instead. However, we could change this function as well, WDYT?

sohkai replied:
I'm definitely thinking making it implicit here sounds like a better abstraction; a `KillSwitch` turning itself off could result in some weird nasty business (although again, right now nothing in `KillSwitch` is `killSwitchProtected` so nothing would happen, which is also quite awkward).

facuspagnuolo commented on contracts/apps/AragonApp.sol:
```diff
@@ -57,6 +64,36 @@ contract AragonApp is AppStorage, Autopetrified, VaultRecoverable, ReentrancyGua
+    * @dev Check whether a call to the current app can be executed or not based on the kill-switch settings
+    * @return Boolean indicating whether the call could be executed or not
+    */
+    function canExecuteCall() public view returns (bool) {
```
I like the idea, I'll go with `isCallEnabled`

facuspagnuolo commented on contracts/kernel/IKernel.sol:
```diff
@@ -20,4 +20,5 @@ contract IKernel is IKernelEvents, IVaultRecoverable {
 
     function setApp(bytes32 namespace, bytes32 appId, address app) public;
     function getApp(bytes32 namespace, bytes32 appId) public view returns (address);
+    function shouldDenyCallingContract(bytes32 appId, address _instance) public returns (bool);
```
Sure, I like the idea :)

facuspagnuolo commented on contracts/factory/DAOFactory.sol:
```diff
@@ -1,33 +1,46 @@
     IACL public baseACL;
-    EVMScriptRegistryFactory public regFactory;
+    KillSwitch public baseKillSwitch;
+    EVMScriptRegistryFactory public scriptsRegistryFactory;
```
I see, I'll roll it back

facuspagnuolo commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,142 @@
+        return deniedBaseImplementations[_base];
+    }
+
+    function checkSeverityIssuesFor(bytes32 _appId, address _base) public view returns (bool) {
```
I like `hasExceededAllowedSeverity` 👍

facuspagnuolo commented on contracts/kill_switch/KillSwitch.sol:
```diff
@@ -0,0 +1,142 @@
+
+    IIssuesRegistry public defaultIssuesRegistry;
+    mapping (address => bool) internal allowedInstances;
+    mapping (address => bool) internal deniedBaseImplementations;
```
We can, would you change the interface as well? I mean, `isInstanceAllowed` to `isInstanceWhitelisted` for example?

facuspagnuolo replied:
Nvm, I already changed them. I think it sounds more robust to refer them as whitelist/blacklist instead 👍

facuspagnuolo commented on contracts/apps/AragonApp.sol:
```diff
@@ -57,6 +64,36 @@ contract AragonApp is AppStorage, Autopetrified, VaultRecoverable, ReentrancyGua
+        bytes memory callData = abi.encodeWithSelector(selector, appId(), address(this));
+        bool success = address(_kernel).call(callData);
+
+        // if the call to `kernel.shouldDenyCallingApp` reverts (using an old Kernel) we consider that
```
Let's add it to the `canPerform` method 👍

facuspagnuolo replied:
Another thing to have in mind is that we will be hiding too much information behind the  `APP_AUTH_FAILED` error message, it could be somehow misleading.

sohkai replied:
Yeah, the thinking behind putting it inside `Kernel.hasPermission()` was to streamline the API and reduce a bit of the gas overhead, as I imagine it would be _very_ common to put the `auth` and `killswitchProtected` modifiers together for most apps.

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+
+  context('when the kernel version does not support kill-switch logic', async () => {
+    before('create DAO factory', async () => {
+      daoFactory = await DAOFactory.new(kernelWithoutKillSwitchBase.address, aclBase.address, ZERO_ADDRESS, registryFactory.address)
```
What do you think about no using the `DAOFactory` in these tests, but installing a `KillSwitch` manually, so that it's really only unit testing the `KillSwitch` (then the `DAOFactory` tests can act as more "integration" type tests).

facuspagnuolo replied:
We could but I think we are covering more contracts and avoiding code duplication if we use it. However, if you prefer to follow that pattern I can refactor them.

sohkai replied:
I'm wondering if it might be worth splitting the tests into `kill-switch/kill_switch.js`, which are mostly unit tests for the exposed `KillSwitch` functionality, and `kernel/kernel_kill_switch.js`, which are then the full suite of integration tests that interact with the various `Kernel`s / mocks.

facuspagnuolo commented on contracts/apps/AragonApp.sol:
```diff
@@ -57,6 +64,36 @@ contract AragonApp is AppStorage, Autopetrified, VaultRecoverable, ReentrancyGua
+        // if not, first ensure the returned value is 32 bytes length
+        uint256 _outputLength;
+        assembly { _outputLength := returndatasize }
+        require(_outputLength == 32, ERROR_UNEXPECTED_KERNEL_RESPONSE);
```
At least for now, we are using this check within a `require` statement, i.e. to raise or not an error. Thus, returning false or reverting here will result in the same effect. And I think that between having more context to know why a call reverted and not, I prefer having it. Do you have in mind some particular scenario where it could be useful swallowing that scenario?

sohkai replied:
If someone were to call this function directly, outside of the modifier (e.g. the app itself or another external contract calling this function to see if it was killswitched or not).

One example is with the Token Manager's [`canForward()`](https://github.com/aragon/aragon-apps/blob/master/apps/token-manager/contracts/TokenManager.sol#L269), where if we don't bake-in the killswitch check at the `EVMScriptRunner` level, we'd have to do a query to `canExecuteCall()` there.

I'd also like to make `canPerform()` and `canExecuteCall()` as consistent as possible, so users interact with them using the same error handling model. `canPerform()` masks incorrect oracle param responses with `false`, rather than reverting.

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+    SET_HIGHEST_ALLOWED_SEVERITY_ROLE = await killSwitchBase.SET_HIGHEST_ALLOWED_SEVERITY_ROLE()
+  })
+
+  context('when the kernel version does not support kill-switch logic', async () => {
```
Doesn't this set of tests check a `KillSwitch`-compatible Kernel with no `KillSwitch` installed? I would say `KernelWithoutKillSwitchMock` is the one that best mimics the behaviour of old Kernels.

facuspagnuolo replied:
Maybe the context message is misleading, but it is actually testing an "old kernel" version, that's why I mention that it "does not support kill-switch logic". Actually right below is a `beforeEach` statement deploying an "old" kernel using `KernelWithoutKillSwitchMock` (see [here](https://github.com/aragon/aragonOS/pull/518/files/ce7a29bf9d359abeab95c9a6b9075fe90ed8722d#diff-20fa58b7946252b898fd2dc455170cf6R56))

sohkai replied:
🙏 I might've gotten confused with the context below, this test looks good!

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+  })
+
+  context('when the kernel version does support kill-switch logic', async () => {
+    const SAMPLE_APP_ID = '0x1236000000000000000000000000000000000000000000000000000000000000'
```
Perhaps this constant could be lifted up, I notice we use a similar one in https://github.com/aragon/aragonOS/pull/518/files#diff-20fa58b7946252b898fd2dc455170cf6R65.

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+                  await killSwitch.setDeniedBaseImplementation(appBase.address, true, { from: owner })
+                })
+
+                itExecutesTheCall()
```
Could we add a comment that allowing the instance has higher precedence than denying the base implementation?

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+            })
+
+            context('when the instance being called is not marked as allowed', () => {
+              beforeEach('dot not allow instance', async () => {
```
```suggestion
              beforeEach('do not allow instance', async () => {
```

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+                  await killSwitch.setDeniedBaseImplementation(appBase.address, true, { from: owner })
+                })
+
+                itExecutesTheCall()
```
Shouldn't this revert?

facuspagnuolo replied:
Note that the `context` for this test is `"when the kernel is initialized with a failing kill-switch implementation"`. We decided that if the kill-switch was failing then we will allow calling the app being queried, it would be the same case as an "old-version" kernel

sohkai replied:
:+1: Ahh yes, missed that this was under the "failing kill-switch" context!

facuspagnuolo commented on contracts/factory/DAOFactory.sol:
```diff
@@ -1,33 +1,46 @@
+            baseKillSwitch = _baseKillSwitch;
         }
 
         baseKernel = _baseKernel;
```
The problem is that you won't notice any further changes in the contracts that implements those IDs in the future, we could cache it for a single transaction, but caching it "forever" seems a lot to me. Note that caching it for a single transaction will result in a lot of code duplication, it is the same trade-off we will always face, code simplicity vs. gas optimization. In this case I opted for the first one since it is a single call per DAO, but I'll try to optimize its costs as much as possible.

sohkai replied:
> you won't notice any further changes in the contracts that implement those IDs

Yes, but in the case of `DAOFactory`, there's no way to change the base implementations we get of the `Kernel`, `ACL`, and `KillSwitch`. Outside of funny business with CREATE2, these IDs shouldn't change.

If we want to handle the possible CREATE2 scenario where these bases are replaced, then let's keep the code simpler and just query them when we need them 👍

facuspagnuolo replied:
Anyway, I refactored it in a way we query roles only once per new DAO

facuspagnuolo commented on contracts/factory/DAOFactory.sol:
```diff
@@ -40,38 +53,114 @@ contract DAOFactory {
+
+        _acl.grantPermission(_to, _acl, createPermissionsRole);
+        _acl.setPermissionManager(_to, _acl, createPermissionsRole);
     }
```
Actually we are granting that permission only for the kill-switch case.

facuspagnuolo replied:
As said above, I think we can optimize it a bit but contexts are different, we are granting those permissions to different entities

facuspagnuolo commented:
Thanks for the review @sohkai! I already addressed/answered all your comments. Feel free to take another look if you want.

sohkai commented on test/contracts/kill_switch/kill_switch.test.js:
```diff
@@ -0,0 +1,963 @@
+              it('returns true', async () => {
+                assert.isTrue(await killSwitch.checkSeverityIssuesFor(SAMPLE_APP_ID, appBase.address))
+              })
+            })
```
Perhaps we can also write a test or two here that checks that `checkSeverityIssuesFor()` reflects the latest issue registry set if the specific app gets set to use a different issues registry (or the `KillSwitch` sets a different default registry).

sohkai commented on contracts/apps/AragonApp.sol:
```diff
@@ -47,7 +41,7 @@ contract AragonApp is AppStorage, Autopetrified, VaultRecoverable, ReentrancyGua
-        if (!hasInitialized()) {
+        if (!isCallEnabled()) {
             return false;
         }
```
We can move the `address(linkedKernel) == address(0)` check to be above this, as `isCallEnabled()` will always fail if the kernel doesn't exist. It may also save a _tiny_ bit of gas to inline the two conditions with an `||` 🤷‍♂

sohkai commented on contracts/kernel/Kernel.sol:
```diff
@@ -164,11 +164,13 @@ contract Kernel is IKernel, KernelStorage, KernelAppIds, KernelNamespaceConstant
-    * @dev Tells whether a call to an instance of an app should be denied or not based on the kill-switch settings
+    * @dev Tells whether a call to an instance of an app should be denied or not based on the kill-switch settings.
+    *      Note that we don't need to perform an initialization check since having or not a kill-switch installed
+    *      implicitly means that.
```
English nitpick: would rephrase to

```
Initialization check is implicitly provided by the KillSwitch's existence, as apps can only be installed after initialization
```

sohkai commented on contracts/kill-switch/KillSwitch.sol:
```diff
@@ -0,0 +1,142 @@
+    function hasExceededAllowedSeverity(bytes32 _appId, address _base) public view returns (bool) {
+        IIssuesRegistry.Severity severityFound = getIssuesRegistry(_appId).getSeverityFor(_base);
+        IIssuesRegistry.Severity highestAllowedSeverity = getHighestAllowedSeverity(_appId);
+        return highestAllowedSeverity >= severityFound;
```
Nitpick: this should probably be `highestAllowedSeverity < severityFound` (and `shouldDenyCallingApp()` should return true if this function is true), to stay true to this function's name (**exceeding** allowed severity).

Alternatively we could rename this to `isWithinAllowedSeverity()` to keep the logic the same.

facuspagnuolo replied:
Good catch, I completely forgot that after renaming it :)

sohkai commented on test/contracts/kill-switch/kill_switch.js:
```diff
@@ -0,0 +1,972 @@
+          const { receipt: { cumulativeGasUsed: gasUsedWithoutKillSwitch } } = await app.writeWithoutKillSwitch(10, { from: owner })
+
+          const killSwitchCost = gasUsedWithKillSwitch - gasUsedWithoutKillSwitch
+          assert(killSwitchCost <= 27000, 'kill switch should overload ~27k of gas')
```
Nitpick: wondering if there was a reason to not use [`assert.isAtMost()`](https://www.chaijs.com/api/assert/#method_isatmost) here, so the tests would log the actual gas used on failure?

facuspagnuolo commented:
The coverage task is failing because there are some tests that fail but only when measuring coverage. I've trying to debug it locally but it was completely cumbersome and I couldn't get to know what was going on. After that, I tried running `solidity-coverage` on using `ganache-cli` instead of `testrpc-sc` using [frangio's fork](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/v2.3.0-rc.3/package.json#L57) and it passed. Definitely `solidity-coverage` is an outdated tool that has already been giving us several headaches 😕
